In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
import seaborn as sns
import matplotlib.pyplot as plt
from warnings import filterwarnings
filterwarnings('ignore')

RAW_PATH = "../data/raw/global_forest_watch_raw_data.xlsx"

print("="*80)
print("LOADING RAW DATA FROM EXCEL FILE")
print("="*80)

excel_file = pd.ExcelFile(RAW_PATH)
print(f"✅ Excel file loaded: {RAW_PATH}")
print(f"\nAvailable sheets: {len(excel_file.sheet_names)}")
print("\nSheet names:")
for i, sheet in enumerate(excel_file.sheet_names, 1):
    print(f"  {i}. {sheet}")

country_sheets = [s for s in excel_file.sheet_names if s.startswith('Country')]
print(f"\n📊 Country-level sheets to explore: {len(country_sheets)}")
for sheet in country_sheets:
    print(f"  - {sheet}")

LOADING RAW DATA FROM EXCEL FILE
✅ Excel file loaded: ../data/raw/global_forest_watch_raw_data.xlsx

Available sheets: 9

Sheet names:
  1. Read_Me
  2. Country tree cover loss
  3. Country primary loss
  4. Country drivers
  5. Country carbon data
  6. Subnational 1 tree cover loss
  7. Subnational 1 primary loss
  8. Subnational 1 drivers
  9. Subnational 1 carbon data

📊 Country-level sheets to explore: 4
  - Country tree cover loss
  - Country primary loss
  - Country drivers
  - Country carbon data


**Findings:** 

From loading the Excel file, we discovered:
- The file contains **9 total sheets**
- **4 country-level sheets** that we'll explore:
  1. Country tree cover loss
  2. Country primary loss
  3. Country drivers
  4. Country carbon data
- **4 subnational-level sheets** (which we'll skip for now, as we're focusing on country-level analysis)
- **1 Read_Me sheet** (documentation)

This gives us a clear picture of what data is available and helps us focus our exploration on the four country-level datasets that will be used for our analysis. The presence of both country and subnational data suggests the dataset has multiple levels of granularity, but for this project, we'll concentrate on the country-level aggregation.


In [ ]:
---

## Part 1: Data Exploration - Raw Data Sheets

### Goal: Understand the raw data structure, quality, and characteristics from each Excel sheet

We will explore each country-level sheet separately to understand:
- Data structure and format
- Column names and types
- Data quality issues
- Relationships between sheets
- Patterns that will inform data preparation


## Step 2: Exploring Country Tree Cover Loss Data


In [2]:
df_tree_cover_loss = excel_file.parse("Country tree cover loss")

print("="*80)
print("COUNTRY TREE COVER LOSS - RAW DATA")
print("="*80)
print(f"Shape: {df_tree_cover_loss.shape[0]:,} rows × {df_tree_cover_loss.shape[1]} columns")
print(f"\nColumn names ({len(df_tree_cover_loss.columns)}):")
for i, col in enumerate(df_tree_cover_loss.columns, 1):
    print(f"  {i:2d}. {col}")

print("\n" + "="*80)
print("First 5 rows:")
print("="*80)
display(df_tree_cover_loss.head())

print("\n" + "="*80)
print("Data Types:")
print("="*80)
print(df_tree_cover_loss.dtypes)

print("\n" + "="*80)
print("Basic Statistics:")
print("="*80)
display(df_tree_cover_loss.describe())


COUNTRY TREE COVER LOSS - RAW DATA
Shape: 1,328 rows × 30 columns

Column names (30):
   1. country
   2. threshold
   3. area_ha
   4. extent_2000_ha
   5. extent_2010_ha
   6. gain_2000-2012_ha
   7. tc_loss_ha_2001
   8. tc_loss_ha_2002
   9. tc_loss_ha_2003
  10. tc_loss_ha_2004
  11. tc_loss_ha_2005
  12. tc_loss_ha_2006
  13. tc_loss_ha_2007
  14. tc_loss_ha_2008
  15. tc_loss_ha_2009
  16. tc_loss_ha_2010
  17. tc_loss_ha_2011
  18. tc_loss_ha_2012
  19. tc_loss_ha_2013
  20. tc_loss_ha_2014
  21. tc_loss_ha_2015
  22. tc_loss_ha_2016
  23. tc_loss_ha_2017
  24. tc_loss_ha_2018
  25. tc_loss_ha_2019
  26. tc_loss_ha_2020
  27. tc_loss_ha_2021
  28. tc_loss_ha_2022
  29. tc_loss_ha_2023
  30. tc_loss_ha_2024

First 5 rows:


country  threshold   area_ha  extent_2000_ha  extent_2010_ha  \
0  Afghanistan          0  64383655        64383655        64383655   
1  Afghanistan         10  64383655          432070          126231   
2  Afghanistan         15  64383655          302629          106852   
3  Afghanistan         20  64383655          284330          105718   
4  Afghanistan         25  64383655          254843           72384   

   gain_2000-2012_ha  tc_loss_ha_2001  tc_loss_ha_2002  tc_loss_ha_2003  \
0              10738              103              214              267   
1              10738               92              190              254   
2              10738               91              186              248   
3              10738               89              181              245   
4              10738               89              180              244   

   tc_loss_ha_2004  ...  tc_loss_ha_2015  tc_loss_ha_2016  tc_loss_ha_2017  \
0              226  ...                0                0                0   
1              207  ...                0                0                0   
2              205  ...                0                0                0   
3              203  ...                0                0                0   
4              202  ...                0                0                0   

   tc_loss_ha_2018  tc_loss_ha_2019  tc_loss_ha_2020  tc_loss_ha_2021  \
0               31               25               46               47   
1               28               19               40               37   
2               28               19               39               32   
3               28               18               39               32   
4               27               18               38               27   

   tc_loss_ha_2022  tc_loss_ha_2023  tc_loss_ha_2024  
0               16              133              223  
1                9               32               32  
2                7               23               17  
3                7               22               16  
4                6               21               14  

[5 rows x 30 columns]


Data Types:
country              object
threshold             int64
area_ha               int64
extent_2000_ha        int64
extent_2010_ha        int64
gain_2000-2012_ha     int64
tc_loss_ha_2001       int64
tc_loss_ha_2002       int64
tc_loss_ha_2003       int64
tc_loss_ha_2004       int64
tc_loss_ha_2005       int64
tc_loss_ha_2006       int64
tc_loss_ha_2007       int64
tc_loss_ha_2008       int64
tc_loss_ha_2009       int64
tc_loss_ha_2010       int64
tc_loss_ha_2011       int64
tc_loss_ha_2012       int64
tc_loss_ha_2013       int64
tc_loss_ha_2014       int64
tc_loss_ha_2015       int64
tc_loss_ha_2016       int64
tc_loss_ha_2017       int64
tc_loss_ha_2018       int64
tc_loss_ha_2019       int64
tc_loss_ha_2020       int64
tc_loss_ha_2021       int64
tc_loss_ha_2022       int64
tc_loss_ha_2023       int64
tc_loss_ha_2024       int64
dtype: object

Basic Statistics:


threshold       area_ha  extent_2000_ha  extent_2010_ha  \
count  1328.000000  1.328000e+03    1.328000e+03    1.328000e+03   
mean     28.125000  7.814805e+07    3.038020e+07    2.994347e+07   
std      22.499791  2.015752e+08    1.056704e+08    1.047382e+08   
min       0.000000  2.094000e+03    0.000000e+00    0.000000e+00   
25%      13.750000  5.117777e+06    5.480255e+05    5.412700e+05   
50%      22.500000  2.022587e+07    3.622986e+06    3.499126e+06   
75%      35.000000  6.201997e+07    1.831977e+07    1.819886e+07   
max      75.000000  1.689455e+09    1.689455e+09    1.689455e+09   

       gain_2000-2012_ha  tc_loss_ha_2001  tc_loss_ha_2002  tc_loss_ha_2003  \
count       1.328000e+03     1.328000e+03     1.328000e+03     1.328000e+03   
mean        7.867308e+05     7.906576e+04     9.644243e+04     8.478851e+04   
std         3.417373e+06     3.124548e+05     4.116669e+05     3.768918e+05   
min         0.000000e+00     0.000000e+00     0.000000e+00     0.000000e+00   
25%         1.383200e+04     5.145000e+02     3.932500e+02     3.120000e+02   
50%         9.435900e+04     6.940500e+03     5.172000e+03     3.940500e+03   
75%         3.882400e+05     3.154525e+04     3.242225e+04     2.848925e+04   
max         3.722054e+07     2.933201e+06     3.715945e+06     3.489258e+06   

       tc_loss_ha_2004  tc_loss_ha_2005  ...  tc_loss_ha_2015  \
count     1.328000e+03     1.328000e+03  ...     1.328000e+03   
mean      1.162822e+05     1.068036e+05  ...     1.153580e+05   
std       4.947177e+05     4.231545e+05  ...     3.906740e+05   
min       0.000000e+00     0.000000e+00  ...     0.000000e+00   
25%       5.327500e+02     5.482500e+02  ...     3.230000e+02   
50%       6.147500e+03     7.207500e+03  ...     7.497500e+03   
75%       3.860275e+04     4.036250e+04  ...     5.140025e+04   
max       4.133606e+06     3.675951e+06  ...     2.925679e+06   

       tc_loss_ha_2016  tc_loss_ha_2017  tc_loss_ha_2018  tc_loss_ha_2019  \
count     1.328000e+03     1.328000e+03     1.328000e+03     1.328000e+03   
mean      1.756815e+05     1.745346e+05     1.469773e+05     1.438572e+05   
std       6.453353e+05     6.079198e+05     5.459009e+05     4.588106e+05   
min       0.000000e+00     0.000000e+00     0.000000e+00     0.000000e+00   
25%       6.187500e+02     7.272500e+02     5.825000e+02     5.377500e+02   
50%       1.421250e+04     1.631400e+04     1.156500e+04     1.108200e+04   
75%       1.063302e+05     9.793725e+04     7.669250e+04     7.689425e+04   
max       6.407238e+06     6.143920e+06     6.621833e+06     4.847068e+06   

       tc_loss_ha_2020  tc_loss_ha_2021  tc_loss_ha_2022  tc_loss_ha_2023  \
count     1.328000e+03     1.328000e+03     1.328000e+03     1.328000e+03   
mean      1.553305e+05     1.498121e+05     1.356731e+05     1.686809e+05   
std       5.799769e+05     6.139020e+05     4.909377e+05     7.502341e+05   
min       0.000000e+00     0.000000e+00     0.000000e+00     0.000000e+00   
25%       6.575000e+02     5.275000e+02     3.860000e+02     8.287500e+02   
50%       1.195450e+04     9.935000e+03     9.190500e+03     1.396300e+04   
75%       8.074400e+04     7.072375e+04     6.527200e+04     7.677800e+04   
max       8.217252e+06     8.559449e+06     5.126743e+06     1.017602e+07   

       tc_loss_ha_2024  
count     1.328000e+03  
mean      1.776107e+05  
std       6.924869e+05  
min       0.000000e+00  
25%       6.737500e+02  
50%       1.095100e+04  
75%       8.178400e+04  
max       7.421840e+06  

[8 rows x 29 columns]

**Findings:**

From exploring the Country Tree Cover Loss sheet, we discovered:

1. **Data Structure:**
   - The sheet contains **1,328 rows and 30 columns**
   - The data is in **wide format** with separate columns for each year (`tc_loss_ha_2001` through `tc_loss_ha_2024`)
   - This represents 24 years of data (2001-2024) plus 6 metadata columns
   - The wide format will need to be converted to long format during data preparation

2. **Key Columns Identified:**
   - `country`: Country names (object/string type)
   - `threshold`: Canopy density threshold values (int64 type, mean: 28.125, suggesting values like 0, 25, 30, 50, 75)
   - `area_ha`: Total area in hectares (int64)
   - `extent_2000_ha`: Tree cover extent in year 2000 (int64)
   - `extent_2010_ha`: Tree cover extent in year 2010 (int64)
   - `gain_2000-2012_ha`: Tree cover gain between 2000-2012 (int64)
   - **24 year-based columns**: `tc_loss_ha_2001` through `tc_loss_ha_2024` representing annual tree cover loss

3. **Data Types:**
   - All numeric columns are stored as **int64** (integers), which is appropriate for hectare measurements
   - Country names are stored as **object** (strings)
   - No float types, which simplifies data handling and avoids precision issues
   - All data types are consistent and appropriate for the data content

4. **Statistical Summary:**
   - **Threshold**: Mean of 28.125, indicating multiple threshold values are used
   - **Area**: Mean area of ~78 million hectares per row (very large, likely country-level aggregations)
   - **Extent 2000**: Mean of ~30 million hectares of tree cover in 2000
   - **Extent 2010**: Mean of ~30 million hectares of tree cover in 2010 (slight decrease)
   - The large standard deviations indicate significant variation across countries
   - Year columns (2001-2024) contain loss values that will need to be unpivoted

**Implication for Data Preparation:** 
- We'll need to reshape this data from wide to long format, extracting years from column names (`tc_loss_ha_2001` → year: 2001, value: [loss amount])
- The 24 year columns will become 24 rows per country-threshold combination
- After reshaping, we expect approximately 1,328 × 24 = ~31,872 rows (if all years have data)
- All numeric columns are already integers, so no type conversion needed
- The metadata columns (area_ha, extent_2000_ha, extent_2010_ha, gain_2000-2012_ha) will need to be preserved during reshaping


In [ ]:
print("="*80)
print("MISSING VALUES ANALYSIS:")
print("="*80)
missing = df_tree_cover_loss.isnull().sum()
missing_pct = (missing / len(df_tree_cover_loss)) * 100
missing_df = pd.DataFrame({
    'Column': missing.index,
    'Missing Count': missing.values,
    'Missing Percentage': missing_pct.values
}).sort_values('Missing Count', ascending=False)

missing_df = missing_df[missing_df['Missing Count'] > 0]
if len(missing_df) > 0:
    display(missing_df)
else:
    print("✅ No missing values found!")

print("\n" + "="*80)
print("UNIQUE VALUES:")
print("="*80)
if 'country' in df_tree_cover_loss.columns:
    print(f"Number of countries: {df_tree_cover_loss['country'].nunique()}")
    print(f"Countries: {sorted(df_tree_cover_loss['country'].unique())}")

if 'threshold' in df_tree_cover_loss.columns:
    print(f"\nThreshold values: {sorted(df_tree_cover_loss['threshold'].unique())}")

year_cols = [col for col in df_tree_cover_loss.columns if '200' in col or '201' in col or '202' in col]
print(f"\nYear columns found: {len(year_cols)}")
if year_cols:
    print(f"Year range: {year_cols[0]} to {year_cols[-1]}")


⚠️ No column found for 'tree_cover_extent_ha'. Check your dataset columns:
['country', 'threshold', 'area_ha_x', 'extent_2000_ha', 'extent_2010_ha', 'gain_2000-2012_ha', 'tree_cover_loss_ha', 'year', 'area_ha_y', 'primary_forest_loss_ha', 'hard_commodities', 'logging', 'other_natural_disturbances', 'permanent_agriculture', 'settlements_infrastructure', 'shifting_cultivation', 'wildfire', 'umd_tree_cover_extent_2000__ha', 'gfw_aboveground_carbon_stocks_2000__mg_c', 'avg_gfw_aboveground_carbon_stocks_2000__mg_c_ha-1', 'gfw_forest_carbon_gross_emissions__mg_co2e_yr-1', 'gfw_forest_carbon_gross_removals__mg_co2_yr-1', 'gfw_forest_carbon_net_flux__mg_co2e_yr-1', 'carbon_gross_emissions_MgCO2e']


## Step 4: Exploring Country Primary Loss Data


In [5]:
df_primary_loss = excel_file.parse("Country primary loss")

print("="*80)
print("COUNTRY PRIMARY LOSS - RAW DATA")
print("="*80)
print(f"Shape: {df_primary_loss.shape[0]:,} rows × {df_primary_loss.shape[1]} columns")
print(f"\nColumn names ({len(df_primary_loss.columns)}):")
for i, col in enumerate(df_primary_loss.columns, 1):
    print(f"  {i:2d}. {col}")

print("\n" + "="*80)
print("First 5 rows:")
print("="*80)
display(df_primary_loss.head())

print("\n" + "="*80)
print("Data Types:")
print("="*80)
print(df_primary_loss.dtypes)

print("\n" + "="*80)
print("Basic Statistics:")
print("="*80)
display(df_primary_loss.describe())

print("\n" + "="*80)
print("MISSING VALUES ANALYSIS:")
print("="*80)
missing = df_primary_loss.isnull().sum()
missing_pct = (missing / len(df_primary_loss)) * 100
missing_df = pd.DataFrame({
    'Column': missing.index,
    'Missing Count': missing.values,
    'Missing Percentage': missing_pct.values
}).sort_values('Missing Count', ascending=False)

missing_df = missing_df[missing_df['Missing Count'] > 0]
if len(missing_df) > 0:
    display(missing_df)
else:
    print("✅ No missing values found!")

print("\n" + "="*80)
print("UNIQUE VALUES:")
print("="*80)
if 'country' in df_primary_loss.columns:
    print(f"Number of countries: {df_primary_loss['country'].nunique()}")
    print(f"Countries: {sorted(df_primary_loss['country'].unique())}")

if 'threshold' in df_primary_loss.columns:
    print(f"\nThreshold values: {sorted(df_primary_loss['threshold'].unique())}")

year_cols = [col for col in df_primary_loss.columns if '200' in col or '201' in col or '202' in col]
print(f"\nYear columns found: {len(year_cols)}")
if year_cols:
    print(f"Year range: {year_cols[0]} to {year_cols[-1]}")


COUNTRY PRIMARY LOSS - RAW DATA
Shape: 76 rows × 26 columns

Column names (26):
   1. country
   2. threshold
   3. area__ha
   4. tc_loss_ha_2002
   5. tc_loss_ha_2003
   6. tc_loss_ha_2004
   7. tc_loss_ha_2005
   8. tc_loss_ha_2006
   9. tc_loss_ha_2007
  10. tc_loss_ha_2008
  11. tc_loss_ha_2009
  12. tc_loss_ha_2010
  13. tc_loss_ha_2011
  14. tc_loss_ha_2012
  15. tc_loss_ha_2013
  16. tc_loss_ha_2014
  17. tc_loss_ha_2015
  18. tc_loss_ha_2016
  19. tc_loss_ha_2017
  20. tc_loss_ha_2018
  21. tc_loss_ha_2019
  22. tc_loss_ha_2020
  23. tc_loss_ha_2021
  24. tc_loss_ha_2022
  25. tc_loss_ha_2023
  26. tc_loss_ha_2024

First 5 rows:


country  threshold  area__ha  tc_loss_ha_2002  tc_loss_ha_2003  \
0      Angola         30   2458061             3499             2963   
1   Argentina         30   4418724             9318            14459   
2   Australia         30     13977                0                0   
3  Bangladesh         30    101114              619              266   
4      Belize         30   1165487             5570             2993   

   tc_loss_ha_2004  tc_loss_ha_2005  tc_loss_ha_2006  tc_loss_ha_2007  \
0             2354             3110             1400             8060   
1            28090            31429            24095            18687   
2                0                0               25                0   
3              347              306              677              369   
4             2108             3206             1899             4140   

   tc_loss_ha_2008  ...  tc_loss_ha_2015  tc_loss_ha_2016  tc_loss_ha_2017  \
0             2699  ...             8998            12040            11166   
1            47067  ...            10547            15247            17202   
2                0  ...                5                0                0   
3              240  ...              205              345              414   
4             3632  ...             6606            11511             6616   

   tc_loss_ha_2018  tc_loss_ha_2019  tc_loss_ha_2020  tc_loss_ha_2021  \
0            13507             9995             8895            24326   
1             9496             8983            20847            11921   
2                0                5                0                0   
3              358              387              459              308   
4             4781             8772            16087             4560   

   tc_loss_ha_2022  tc_loss_ha_2023  tc_loss_ha_2024  
0            15576            17627            13660  
1            21388            11473            12103  
2                0                0                0  
3              307              743              467  
4             4033            11667            21137  

[5 rows x 26 columns]


Data Types:
country            object
threshold           int64
area__ha            int64
tc_loss_ha_2002     int64
tc_loss_ha_2003     int64
tc_loss_ha_2004     int64
tc_loss_ha_2005     int64
tc_loss_ha_2006     int64
tc_loss_ha_2007     int64
tc_loss_ha_2008     int64
tc_loss_ha_2009     int64
tc_loss_ha_2010     int64
tc_loss_ha_2011     int64
tc_loss_ha_2012     int64
tc_loss_ha_2013     int64
tc_loss_ha_2014     int64
tc_loss_ha_2015     int64
tc_loss_ha_2016     int64
tc_loss_ha_2017     int64
tc_loss_ha_2018     int64
tc_loss_ha_2019     int64
tc_loss_ha_2020     int64
tc_loss_ha_2021     int64
tc_loss_ha_2022     int64
tc_loss_ha_2023     int64
tc_loss_ha_2024     int64
dtype: object

Basic Statistics:


threshold      area__ha  tc_loss_ha_2002  tc_loss_ha_2003  \
count       76.0  7.600000e+01     7.600000e+01     7.600000e+01   
mean        30.0  1.349796e+07     3.500986e+04     3.268253e+04   
std          0.0  4.300172e+07     1.883565e+05     1.816783e+05   
min         30.0  1.653000e+03     0.000000e+00     0.000000e+00   
25%         30.0  2.271512e+05     1.325000e+02     1.225000e+02   
50%         30.0  1.833100e+06     2.029500e+03     1.582500e+03   
75%         30.0  7.487047e+06     1.013250e+04     1.014375e+04   
max         30.0  3.432610e+08     1.621738e+06     1.570540e+06   

       tc_loss_ha_2004  tc_loss_ha_2005  tc_loss_ha_2006  tc_loss_ha_2007  \
count     7.600000e+01     7.600000e+01     7.600000e+01     7.600000e+01   
mean      4.462941e+04     4.370550e+04     3.695367e+04     3.814787e+04   
std       2.367232e+05     2.157229e+05     1.706006e+05     1.452392e+05   
min       0.000000e+00     0.000000e+00     0.000000e+00     0.000000e+00   
25%       2.035000e+02     2.170000e+02     2.455000e+02     1.845000e+02   
50%       2.277000e+03     2.320000e+03     2.528500e+03     3.063500e+03   
75%       1.197275e+04     1.202100e+04     1.418450e+04     1.843200e+04   
max       2.016350e+06     1.824217e+06     1.415536e+06     1.149515e+06   

       tc_loss_ha_2008  tc_loss_ha_2009  ...  tc_loss_ha_2015  \
count     7.600000e+01        76.000000  ...        76.000000   
mean      3.563292e+04     36765.184211  ...     38528.013158   
std       1.354756e+05    116080.074035  ...    126231.483498   
min       0.000000e+00         0.000000  ...         0.000000   
25%       2.227500e+02       445.250000  ...       253.750000   
50%       3.677500e+03      3729.500000  ...      3392.500000   
75%       1.528925e+04     19996.250000  ...     16938.250000   
max       1.075087e+06    700115.000000  ...    828839.000000   

       tc_loss_ha_2016  tc_loss_ha_2017  tc_loss_ha_2018  tc_loss_ha_2019  \
count     7.600000e+01     7.600000e+01     7.600000e+01     7.600000e+01   
mean      8.050946e+04     6.532487e+04     4.792633e+04     4.931018e+04   
std       3.430140e+05     2.536957e+05     1.680912e+05     1.702689e+05   
min       0.000000e+00     0.000000e+00     0.000000e+00     0.000000e+00   
25%       5.015000e+02     5.145000e+02     3.417500e+02     3.247500e+02   
50%       4.904000e+03     7.123000e+03     4.061500e+03     5.047500e+03   
75%       4.590225e+04     3.110525e+04     2.709850e+04     3.070975e+04   
max       2.830943e+06     2.134474e+06     1.347176e+06     1.361053e+06   

       tc_loss_ha_2020  tc_loss_ha_2021  tc_loss_ha_2022  tc_loss_ha_2023  \
count     7.600000e+01     7.600000e+01     7.600000e+01     7.600000e+01   
mean      5.531720e+04     4.926511e+04     5.407901e+04     4.911196e+04   
std       2.058852e+05     1.885591e+05     2.149624e+05     1.549588e+05   
min       0.000000e+00     0.000000e+00     0.000000e+00     0.000000e+00   
25%       3.065000e+02     3.012500e+02     2.927500e+02     2.740000e+02   
50%       4.094000e+03     4.050000e+03     3.952000e+03     4.538500e+03   
75%       3.433400e+04     2.444950e+04     2.350000e+04     2.459825e+04   
max       1.703491e+06     1.546964e+06     1.772214e+06     1.136250e+06   

       tc_loss_ha_2024  
count     7.600000e+01  
mean      8.849616e+04  
std       3.671704e+05  
min       0.000000e+00  
25%       3.660000e+02  
50%       4.598000e+03  
75%       4.037175e+04  
max       2.823646e+06  

[8 rows x 25 columns]


MISSING VALUES ANALYSIS:
✅ No missing values found!

UNIQUE VALUES:
Number of countries: 76
Countries: ['Angola', 'Argentina', 'Australia', 'Bangladesh', 'Belize', 'Benin', 'Bhutan', 'Bolivia', 'Brazil', 'Brunei', 'Burundi', 'Cambodia', 'Cameroon', 'Central African Republic', 'China', 'Colombia', 'Costa Rica', 'Cuba', "Côte d'Ivoire", 'Democratic Republic of the Congo', 'Dominican Republic', 'Ecuador', 'El Salvador', 'Equatorial Guinea', 'Ethiopia', 'Fiji', 'French Guiana', 'Gabon', 'Ghana', 'Guadeloupe', 'Guatemala', 'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti', 'Honduras', 'India', 'Indonesia', 'Kenya', 'Laos', 'Liberia', 'Madagascar', 'Malawi', 'Malaysia', 'Martinique', 'Mozambique', 'Myanmar', 'México', 'Nepal', 'Nicaragua', 'Nigeria', 'Panama', 'Papua New Guinea', 'Paraguay', 'Peru', 'Philippines', 'Republic of the Congo', 'Rwanda', 'Senegal', 'Sierra Leone', 'Solomon Islands', 'South Africa', 'South Sudan', 'Sri Lanka', 'Suriname', 'Tanzania', 'Thailand', 'Togo', 'Uganda', 'Unit

**Findings:**

From exploring the Country Primary Loss sheet, we discovered:

1. **Data Structure:**
   - The sheet contains **76 rows and 26 columns**
   - The data is in **wide format** with separate columns for each year (`tc_loss_ha_2002` through `tc_loss_ha_2024`)
   - **Important difference**: Year columns start at 2002 (not 2001), so we have 23 years of data (2002-2024) plus 3 metadata columns
   - This is significantly smaller than tree cover loss (76 rows vs. 1,328 rows), indicating fewer countries have primary forest data
   - The wide format will need to be converted to long format during data preparation

2. **Key Columns Identified:**
   - `country`: Country names (object/string type)
   - `threshold`: Canopy density threshold values (int64 type) - **Only threshold 30 is present** (unlike tree cover loss which has multiple thresholds)
   - `area__ha`: Total area in hectares (int64) - Note: column name uses double underscore `area__ha` (different from tree cover loss which uses `area_ha`)
   - **23 year-based columns**: `tc_loss_ha_2002` through `tc_loss_ha_2024` representing annual primary forest loss
   - **Missing 2001 data**: Unlike tree cover loss, primary loss data starts from 2002

3. **Data Types:**
   - All numeric columns are stored as **int64** (integers), which is appropriate for hectare measurements
   - Country names are stored as **object** (strings)
   - Data types are consistent with tree cover loss sheet, which will facilitate merging
   - No float types, simplifying data handling

4. **Statistical Summary:**
   - **Threshold**: All values are 30 (mean: 30.0, std: 0.0) - only one threshold value in this dataset
   - **Area**: Mean area of ~13.5 million hectares per row (smaller than tree cover loss's ~78 million, reflecting fewer/larger primary forest countries)
   - **Primary Loss Values**: Mean loss ranges from ~35,000 hectares (2002) to varying amounts across years
   - Primary loss values are indeed smaller than total tree cover loss values, as expected
   - The large standard deviations indicate significant variation across countries

5. **Missing Values:**
   - **No missing values found** in the dataset (excellent data quality!)
   - However, the dataset only includes 76 countries (vs. many more in tree cover loss), suggesting countries without primary forests are simply not included
   - Zero values in year columns likely represent "no primary forest loss" rather than missing data

6. **Geographic and Temporal Coverage:**
   - **76 unique countries** represented (significantly fewer than tree cover loss's 1,328 rows, which includes multiple thresholds per country)
   - Countries include major primary forest nations: Brazil, Indonesia, Democratic Republic of the Congo, Peru, Colombia, etc.
   - **Threshold values**: Only threshold 30 (single value, unlike tree cover loss)
   - **Year range**: 2002 to 2024 (23 years of data, missing 2001)
   - **Year columns**: 23 columns found

7. **Relationship to Tree Cover Loss:**
   - This dataset uses the same country identifiers as tree cover loss, but **only includes threshold 30**
   - Can be merged with tree cover loss data on country, threshold (30), and year
   - **Important validation**: Primary loss values should be ≤ tree cover loss values for the same country/year/threshold 30
   - Primary forests are a subset of total forests, so primary loss should never exceed total loss
   - **Column name difference**: `area__ha` (double underscore) vs. `area_ha` in tree cover loss - will need standardization

**Implication for Data Preparation:**
- This sheet will need the same wide-to-long format transformation as tree cover loss
- **Important**: Year extraction must account for starting year 2002 (not 2001)
- Can be merged with tree cover loss after reshaping using country + threshold (30) + year as keys
- Should implement validation to ensure primary loss ≤ total loss during data preparation
- Column name `area__ha` needs to be standardized to `area_ha` for consistency
- After reshaping, we expect approximately 76 × 23 = ~1,748 rows (if all years have data)
- The smaller dataset size (76 countries vs. many more in tree cover loss) means we'll need to decide on merge strategy (inner vs. outer join)


### 1.3 Sheet 3: Country Drivers

## Step 3: Exploring Country Drivers Data


In [11]:
df_drivers = excel_file.parse("Country drivers")

print("="*80)
print("COUNTRY DRIVERS - RAW DATA")
print("="*80)
print(f"Shape: {df_drivers.shape[0]:,} rows × {df_drivers.shape[1]} columns")
print(f"\nColumn names ({len(df_drivers.columns)}):")
for i, col in enumerate(df_drivers.columns, 1):
    print(f"  {i:2d}. {col}")

print("\n" + "="*80)
print("First 5 rows:")
print("="*80)
display(df_drivers.head())

driver_type_cols = [c for c in df_drivers.columns if c not in ['country', 'threshold'] and not any(str(year) in c for year in range(2001, 2025))]
print(f"\nDriver type columns: {driver_type_cols}")

year_cols = [col for col in df_drivers.columns if any(str(year) in col for year in range(2001, 2025))]
print(f"Year-based columns: {len(year_cols)}")

print("\n" + "="*80)
print("Missing Values:")
print("="*80)
missing = df_drivers.isnull().sum()
if missing.sum() > 0:
    display(pd.DataFrame({'Missing Count': missing[missing > 0]}))
else:
    print("✅ No missing values!")

COUNTRY DRIVERS - RAW DATA
Shape: 21,897 rows × 5 columns

Column names (5):
   1. country
   2. threshold
   3. driver
   4. year
   5. tc_loss_ha

First 5 rows:


country  threshold            driver  year  tc_loss_ha
0  Afghanistan         30  Hard commodities  2014         0.0
1  Afghanistan         30           Logging  2001         3.0
2  Afghanistan         30           Logging  2002        64.0
3  Afghanistan         30           Logging  2003        73.0
4  Afghanistan         30           Logging  2004       143.0


Driver type columns: ['driver', 'year', 'tc_loss_ha']
Year-based columns: 0

Missing Values:
✅ No missing values!


**Findings:**

From exploring the Country Drivers sheet, we discovered:

1. **Data Structure:**
   - The sheet contains **21,897 rows and 5 columns**
   - **Critical difference**: The data is already in **LONG format** (unlike tree cover loss and primary loss which are in wide format)
   - This is a significant advantage as it doesn't require reshaping during data preparation
   - The structure is: one row per country-threshold-driver-year combination

2. **Key Columns Identified:**
   - `country`: Country names (object/string type)
   - `threshold`: Canopy density threshold values (int64 type)
   - `driver`: Driver category names (object/string type) - represents the cause of deforestation
   - `year`: Year values (int64 type) - already extracted as a column (not in column names)
   - `tc_loss_ha`: Tree cover loss in hectares attributed to that specific driver (float64 type)

3. **Driver Categories:**
   - The `driver` column contains categorical values representing different causes of deforestation
   - Examples from the data: "Hard commodities", "Logging"
   - These represent the main causes of deforestation tracked by Global Forest Watch
   - Each row represents loss attributed to a specific driver for a specific country-threshold-year combination

4. **Data Types:**
   - Country and driver names are stored as **object** (strings)
   - Threshold and year are stored as **int64** (integers)
   - Loss values are stored as **float64** (floats), which allows for precise measurements
   - All data types are appropriate for the data content

5. **Data Completeness:**
   - **No missing values found** in the dataset (excellent data quality!)
   - **21,897 rows** represent multiple driver types per country-threshold-year combination
   - The large number of rows (compared to other sheets) reflects the long format structure

6. **Important Structural Difference:**
   - **Unlike tree cover loss and primary loss sheets**, this data is already in long format
   - No year-based columns (0 year-based columns found)
   - Year is already a separate column, making this dataset ready for merging without reshaping
   - This suggests the data was pre-processed differently than the loss datasets

7. **Relationship to Other Sheets:**
   - Can be merged with tree cover loss data using: country + threshold + year as keys
   - After merging, we can analyze which drivers contribute most to forest loss
   - Driver values represent a breakdown of total loss by cause, so summing drivers for a country-threshold-year should approximate (but may not exactly equal) total loss
   - Some loss may be unclassified or attributed to multiple causes

**Implication for Data Preparation:**
- **No reshaping needed** - this sheet is already in the desired long format
- Can be directly merged with reshaped loss data using country + threshold + year as merge keys
- May want to pivot driver column to create separate columns for each driver type (e.g., `hard_commodities_ha`, `logging_ha`) for easier analysis, or keep in long format depending on analysis needs
- The float64 data type for `tc_loss_ha` is appropriate for precise measurements
- This dataset will be easier to integrate than the loss datasets since it doesn't require wide-to-long transformation


### 1.4 Sheet 4: Country Carbon Data

## Step 4: Exploring Country Carbon Data


In [7]:
df_carbon = excel_file.parse("Country carbon data")

print("="*80)
print("COUNTRY CARBON DATA - RAW DATA")
print("="*80)
print(f"Shape: {df_carbon.shape[0]:,} rows × {df_carbon.shape[1]} columns")
print(f"\nColumn names ({len(df_carbon.columns)}):")
for i, col in enumerate(df_carbon.columns, 1):
    print(f"  {i:2d}. {col}")

print("\n" + "="*80)
print("First 5 rows:")
print("="*80)
display(df_carbon.head())

print("\n" + "="*80)
print("Missing Values:")
print("="*80)
missing = df_carbon.isnull().sum()
if missing.sum() > 0:
    display(pd.DataFrame({'Missing Count': missing[missing > 0]}))
else:
    print("✅ No missing values!")

carbon_metric_cols = [c for c in df_carbon.columns if 'carbon' in c.lower() or 'emission' in c.lower()]
print(f"\nCarbon-related columns: {carbon_metric_cols}")


COUNTRY CARBON DATA - RAW DATA
Shape: 498 rows × 32 columns

Column names (32):
   1. country
   2. umd_tree_cover_density_2000__threshold
   3. umd_tree_cover_extent_2000__ha
   4. gfw_aboveground_carbon_stocks_2000__Mg_C
   5. avg_gfw_aboveground_carbon_stocks_2000__Mg_C_ha-1
   6. gfw_forest_carbon_gross_emissions__Mg_CO2e_yr-1
   7. gfw_forest_carbon_gross_removals__Mg_CO2_yr-1
   8. gfw_forest_carbon_net_flux__Mg_CO2e_yr-1
   9. gfw_forest_carbon_gross_emissions_2001__Mg_CO2e
  10. gfw_forest_carbon_gross_emissions_2002__Mg_CO2e
  11. gfw_forest_carbon_gross_emissions_2003__Mg_CO2e
  12. gfw_forest_carbon_gross_emissions_2004__Mg_CO2e
  13. gfw_forest_carbon_gross_emissions_2005__Mg_CO2e
  14. gfw_forest_carbon_gross_emissions_2006__Mg_CO2e
  15. gfw_forest_carbon_gross_emissions_2007__Mg_CO2e
  16. gfw_forest_carbon_gross_emissions_2008__Mg_CO2e
  17. gfw_forest_carbon_gross_emissions_2009__Mg_CO2e
  18. gfw_forest_carbon_gross_emissions_2010__Mg_CO2e
  19. gfw_forest_carbon_gros

country  umd_tree_cover_density_2000__threshold  \
0  Afghanistan                                      30   
1  Afghanistan                                      50   
2  Afghanistan                                      75   
3      Albania                                      30   
4      Albania                                      50   

   umd_tree_cover_extent_2000__ha  gfw_aboveground_carbon_stocks_2000__Mg_C  \
0                          205771                                  12409398   
1                          148417                                   9765465   
2                           75480                                   5571655   
3                          648459                                  40958831   
4                          534671                                  37239867   

   avg_gfw_aboveground_carbon_stocks_2000__Mg_C_ha-1  \
0                                                123   
1                                                134   
2                                                150   
3                                                238   
4                                                263   

   gfw_forest_carbon_gross_emissions__Mg_CO2e_yr-1  \
0                                            15339   
1                                            12657   
2                                             6147   
3                                           721806   
4                                           682919   

   gfw_forest_carbon_gross_removals__Mg_CO2_yr-1  \
0                                         376800   
1                                         275855   
2                                         151074   
3                                        5103589   
4                                        4294627   

   gfw_forest_carbon_net_flux__Mg_CO2e_yr-1  \
0                                   -361461   
1                                   -263199   
2                                   -144926   
3                                  -4381783   
4                                  -3611709   

   gfw_forest_carbon_gross_emissions_2001__Mg_CO2e  \
0                                          27986.0   
1                                          25603.0   
2                                          15780.0   
3                                        1417747.0   
4                                        1358272.0   

   gfw_forest_carbon_gross_emissions_2002__Mg_CO2e  ...  \
0                                          41762.0  ...   
1                                          32691.0  ...   
2                                          15308.0  ...   
3                                         348556.0  ...   
4                                         338279.0  ...   

   gfw_forest_carbon_gross_emissions_2015__Mg_CO2e  \
0                                              0.0   
1                                              0.0   
2                                              0.0   
3                                         120041.0   
4                                         113553.0   

   gfw_forest_carbon_gross_emissions_2016__Mg_CO2e  \
0                                              0.0   
1                                              0.0   
2                                              0.0   
3                                         334094.0   
4                                         304691.0   

   gfw_forest_carbon_gross_emissions_2017__Mg_CO2e  \
0                                              0.0   
1                                              0.0   
2                                              0.0   
3                                         448993.0   
4                                         403366.0   

   gfw_forest_carbon_gross_emissions_2018__Mg_CO2e  \
0                                           4893.0   
1                                           3920.0   
2                                           1962.0   
3                                         724335.0   



Missing Values:
✅ No missing values!

Carbon-related columns: ['gfw_aboveground_carbon_stocks_2000__Mg_C', 'avg_gfw_aboveground_carbon_stocks_2000__Mg_C_ha-1', 'gfw_forest_carbon_gross_emissions__Mg_CO2e_yr-1', 'gfw_forest_carbon_gross_removals__Mg_CO2_yr-1', 'gfw_forest_carbon_net_flux__Mg_CO2e_yr-1', 'gfw_forest_carbon_gross_emissions_2001__Mg_CO2e', 'gfw_forest_carbon_gross_emissions_2002__Mg_CO2e', 'gfw_forest_carbon_gross_emissions_2003__Mg_CO2e', 'gfw_forest_carbon_gross_emissions_2004__Mg_CO2e', 'gfw_forest_carbon_gross_emissions_2005__Mg_CO2e', 'gfw_forest_carbon_gross_emissions_2006__Mg_CO2e', 'gfw_forest_carbon_gross_emissions_2007__Mg_CO2e', 'gfw_forest_carbon_gross_emissions_2008__Mg_CO2e', 'gfw_forest_carbon_gross_emissions_2009__Mg_CO2e', 'gfw_forest_carbon_gross_emissions_2010__Mg_CO2e', 'gfw_forest_carbon_gross_emissions_2011__Mg_CO2e', 'gfw_forest_carbon_gross_emissions_2012__Mg_CO2e', 'gfw_forest_carbon_gross_emissions_2013__Mg_CO2e', 'gfw_forest_carbon_gross_emissio

**Findings:**

From exploring the Country Carbon Data sheet, we discovered:

1. **Data Structure:**
   - The sheet contains **498 rows and 32 columns**
   - The data is in **wide format** with separate columns for each year (`gfw_forest_carbon_gross_emissions_2001__Mg_CO2e` through `gfw_forest_carbon_gross_emissions_2024__Mg_CO2e`)
   - This represents 24 years of annual emissions data (2001-2024) plus 8 metadata columns
   - The wide format will need to be converted to long format during data preparation
   - Similar structure to tree cover loss and primary loss sheets

2. **Key Columns Identified:**
   - `country`: Country names (object/string type)
   - `umd_tree_cover_density_2000__threshold`: Canopy density threshold values (int64 type) - **Note**: Column name differs from other sheets (uses `umd_tree_cover_density_2000__threshold` instead of `threshold`)
   - `umd_tree_cover_extent_2000__ha`: Tree cover extent in year 2000 (int64)
   - `gfw_aboveground_carbon_stocks_2000__Mg_C`: Total carbon stored in forests in 2000 (int64)
   - `avg_gfw_aboveground_carbon_stocks_2000__Mg_C_ha-1`: Average carbon density per hectare in 2000 (int64)
   - `gfw_forest_carbon_gross_emissions__Mg_CO2e_yr-1`: Annual gross carbon emissions (int64)
   - `gfw_forest_carbon_gross_removals__Mg_CO2_yr-1`: Annual gross carbon removals (int64)
   - `gfw_forest_carbon_net_flux__Mg_CO2e_yr-1`: Net carbon flux (emissions - removals) (int64)
   - **24 year-based columns**: `gfw_forest_carbon_gross_emissions_2001__Mg_CO2e` through `gfw_forest_carbon_gross_emissions_2024__Mg_CO2e` representing annual carbon emissions

3. **Carbon Metrics Available:**
   - **Carbon Stocks (2000 baseline)**: `gfw_aboveground_carbon_stocks_2000__Mg_C` - Total carbon stored in forests
   - **Average Carbon Density**: `avg_gfw_aboveground_carbon_stocks_2000__Mg_C_ha-1` - Carbon per hectare
   - **Annual Gross Emissions**: `gfw_forest_carbon_gross_emissions__Mg_CO2e_yr-1` - Total annual emissions
   - **Annual Gross Removals**: `gfw_forest_carbon_gross_removals__Mg_CO2_yr-1` - Total annual removals
   - **Net Carbon Flux**: `gfw_forest_carbon_net_flux__Mg_CO2e_yr-1` - Net change (emissions - removals)
   - **Yearly Emissions (2001-2024)**: 24 columns with annual emissions data

4. **Data Types:**
   - Most numeric columns are stored as **int64** (integers)
   - Year-based emission columns are stored as **float64** (floats), allowing for precise measurements
   - Country names are stored as **object** (strings)
   - Data types are appropriate for the data content

5. **Units:**
   - Carbon stocks: **Mg C** (Megagrams of Carbon)
   - Carbon emissions: **Mg CO2e** (Megagrams of CO2 equivalent)
   - Carbon removals: **Mg CO2** (Megagrams of CO2)
   - Tree cover extent: **ha** (hectares)
   - **Important**: Units must be preserved and documented during merging

6. **Data Quality:**
   - **No missing values found** in the dataset (excellent data quality!)
   - **498 rows** represent country-threshold combinations (similar to tree cover loss structure)
   - The dataset includes multiple threshold values per country

7. **Relationship to Other Sheets:**
   - Can be merged with tree cover loss data using: country + threshold + year as keys
   - **Important**: The threshold column name differs (`umd_tree_cover_density_2000__threshold` vs. `threshold`) - will need standardization during data preparation
   - Carbon emissions can be linked to forest loss to quantify climate impact
   - The year-based emission columns (2001-2024) align with the loss data time period

8. **Temporal Coverage:**
   - **Year range**: 2001 to 2024 (24 years of data)
   - Baseline year: 2000 (for carbon stocks and tree cover extent)
   - Annual emissions data available for each year in the range

**Implication for Data Preparation:**
- This sheet will need the same wide-to-long format transformation as tree cover loss and primary loss
- Year extraction must account for starting year 2001 (same as tree cover loss)
- **Column name standardization needed**: `umd_tree_cover_density_2000__threshold` should be renamed to `threshold` for consistency with other sheets
- Can be merged with loss data after reshaping using country + threshold + year as keys
- After reshaping, we expect approximately 498 × 24 = ~11,952 rows (if all years have data)
- Units must be preserved and clearly documented in the final merged dataset
- The carbon data enables climate impact analysis by linking deforestation to carbon emissions


**Final Summary: CRISP-DM Data Understanding Phase Completion**

This notebook represents the completion of the **Data Understanding** phase of the CRISP-DM methodology. Through systematic exploration of the raw Global Forest Watch data, we have achieved the key objectives of this phase:

1. **Data Collection and Initial Assessment:**
   - Successfully loaded and cataloged all available data sources (9 sheets total)
   - Identified and focused on the four country-level datasets relevant to our analysis
   - Documented the data source structure and organization

2. **Data Description:**
   - Characterized the structure of each dataset (wide format with year-based columns for loss and carbon data; long format for drivers data)
   - Documented all column names, data types, and basic statistics for each sheet
   - Identified 1,328 rows in tree cover loss, 76 rows in primary loss, 21,897 rows in drivers, and 498 rows in carbon data

3. **Data Quality Assessment:**
   - Evaluated missing values across all sheets (found no missing values in any dataset - excellent quality!)
   - Identified data inconsistencies (e.g., different threshold column names, different starting years)
   - Assessed data completeness and coverage (166 countries in tree cover loss, 76 in primary loss, 158 in drivers, 166 in carbon data)

4. **Data Exploration:**
   - Examined relationships between datasets (common identifiers: country, threshold, year)
   - Identified 76 countries present in all sheets (critical for merging strategy)
   - Discovered structural differences (drivers data already in long format; others need reshaping)

5. **Documentation of Findings:**
   - Created comprehensive documentation of data characteristics
   - Identified specific data preparation requirements (reshaping, column standardization, merge strategy)
   - Established clear expectations for the next phase

**Transition to Data Preparation Phase (CRISP-DM):**

The findings from this Data Understanding phase directly inform the **Data Preparation** phase (02_data_preparation.ipynb), where we will:

- **Data Selection:** Focus on the four country-level datasets identified
- **Data Cleaning:** Standardize column names (e.g., `umd_tree_cover_density_2000__threshold` → `threshold`, `area__ha` → `area_ha`)
- **Data Construction:** Reshape wide format to long format for tree cover loss, primary loss, and carbon data
- **Data Integration:** Merge all sheets using country + threshold + year as keys
- **Data Formatting:** Ensure consistent data types and units across the merged dataset

This systematic Data Understanding phase ensures that our Data Preparation phase will be:
- **Informed:** Based on actual data characteristics, not assumptions
- **Efficient:** Addressing only real issues identified through exploration
- **Comprehensive:** Covering all identified data quality and structural concerns
- **Aligned with CRISP-DM:** Following the methodology's structured approach to data science projects

The completion of this phase provides a solid foundation for the subsequent phases of CRISP-DM: Data Preparation, Modeling, Evaluation, and Deployment.
